components/cam/src/physics/cam/micro_p3.F90

```diff
@@ -3045,6 +2888,425 @@ subroutine check_values(Qv,T,i,timestepcount,force_abort,source_ind)
     endif
 
   end subroutine check_values
-  !===========================================================================================
+
+  subroutine ice_droplet_collection(rho,t,rhofaci,    &
+  f1pr04,qitot_incld,qc_incld,nitot_incld,nc_incld,    &
+             qccol,nccol,qcshd,ncshdc)
+   
+   !.......................
+   ! collection of droplets
+
+   ! here we multiply rates by air density, air density fallspeed correction
+   ! factor, and collection efficiency since these parameters are not
+   ! included in lookup table calculations
+   ! for T < 273.15, assume collected cloud water is instantly frozen
+   ! note 'f1pr' values are normalized, so we need to multiply by N
+
+
+   implicit none 
+
+   real(rtype), intent(in) :: rho
+   real(rtype), intent(in) :: t 
+   real(rtype), intent(in) :: rhofaci 
+   real(rtype), intent(in) :: f1pr04
+   real(rtype), intent(in) :: qitot_incld 
+   real(rtype), intent(in) :: qc_incld
+   real(rtype), intent(in) :: nitot_incld 
+   real(rtype), intent(in) :: nc_incld 
+
+   
+   real(rtype), intent(out) :: qccol
+   real(rtype), intent(out) :: nccol
+   real(rtype), intent(out) :: qcshd 
+   real(rtype), intent(out) :: ncshdc
+
+   if (qitot_incld .ge.qsmall .and. qc_incld .ge.qsmall) then 
+      if  (t .le.zerodegc) then
+         qccol = rhofaci*f1pr04*qc_incld*eci*rho*nitot_incld
+         nccol = rhofaci*f1pr04*nc_incld*eci*rho*nitot_incld
+      else if (t .gt. zerodegc) then 
+         ! for T > 273.15, assume cloud water is collected and shed as rain drops
+         ! sink for cloud water mass and number, note qcshed is source for rain mass
+         qcshd = rhofaci*f1pr04*qc_incld*eci*rho*nitot_incld
+         nccol = rhofaci*f1pr04*nc_incld*eci*rho*nitot_incld
+         ! source for rain number, assume 1 mm drops are shed
+         ncshdc = qcshd*1.923e+6_rtype
+      end if 
+   end if 
+
+  end subroutine ice_droplet_collection
+
+
+  subroutine ice_drop_collection(rho,t,rhofaci,    &
+  logn0r,f1pr07,f1pr08,qitot_incld,nitot_incld,qr_incld,    &
+  qrcol, nrcol)
+   
+   !....................
+   ! collection of rain
+
+   ! here we multiply rates by air density, air density fallspeed correction
+   ! factor, collection efficiency, and n0r since these parameters are not
+   ! included in lookup table calculations
+
+   ! for T < 273.15, assume all collected rain mass freezes
+   ! note this is a sink for rain mass and number and a source
+   ! for ice mass
+
+   ! note 'f1pr' values are normalized, so we need to multiply by N
+
+   implicit none 
+
+   real(rtype), intent(in) :: rho
+   real(rtype), intent(in) :: t 
+   real(rtype), intent(in) :: rhofaci 
+   real(rtype), intent(in) :: logn0r 
+   real(rtype), intent(in) :: f1pr07 
```

Comment:
`f1pr07   ! collection of rain number by ice` and `f1pr08   ! collection of rain mass by ice`